### PR TITLE
Add admin login

### LINKED ISSUE #101: Create back end management console
Create a login area with authentication.

At the moment, this will be restricted to admin users, so no access control will be required initially.

We may add attendee logins in the future - see #102 (this would include access control implementation).

### PR DESCRIPTION
**Pull request by asgrim**
Added empty dashboard, ability to log in etc.

Fixes #101

### CODE REVIEW

File: config/autoload/routes.global.php [hunk 1]
```diff
@@ -87,5 +91,29 @@
...
+            'allowed_methods' => ['GET', 'POST'],
+        ],
+        [
+            'name' => 'account-dashboard',
+            'path' => '/account/dashboard',
+            'middleware' => [
+                App\Middleware\Authentication::class,
+                App\Action\Account\DashboardAction::class,
+            ],
+            'allowed_methods' => ['GET', 'POST'],
```
Comment (lboynton) [inline config/autoload/routes.global.php]: Is this using POST?

File: src/App/Middleware/AuthenticationFactory.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+use App\Service\Authentication\AuthenticationServiceInterface;
+use Interop\Container\ContainerInterface;
+use Zend\Expressive\Template\TemplateRendererInterface;
+
+/**
+ * @codeCoverageIgnore
+ */
+final class AuthenticationFactory
+{
+    public function __invoke(ContainerInterface $container)
```
Comment (lboynton) [inline src/App/Middleware/AuthenticationFactory.php]: RTH

File: src/App/Service/Authentication/Factory.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * @codeCoverageIgnore
+ */
+class Factory implements FactoryInterface
+{
+    /**
+     * {@inheritdoc}
+     * @throws \Interop\Container\Exception\NotFoundException
+     */
+    public function __invoke(ContainerInterface $container, $requestedName, array $options = null)
```
Comment (lboynton) [inline src/App/Service/Authentication/Factory.php]: RTH

File: src/App/Service/User/FindUserByEmailFactory.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * @codeCoverageIgnore
+ */
+class FindUserByEmailFactory implements FactoryInterface
+{
+    /**
+     * {@inheritdoc}
+     * @throws \Interop\Container\Exception\NotFoundException
+     */
+    public function __invoke(ContainerInterface $container, $requestedName, array $options = null)
```
Comment (lboynton) [inline src/App/Service/User/FindUserByEmailFactory.php]: RTH

File: src/App/Middleware/Authentication.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    ) {
+        $this->authenticationService = $authenticationService;
+        $this->renderer = $renderer;
+    }
+
+    /**
+     * {@inheritdoc}
+     * @throws \InvalidArgumentException
+     */
+    public function __invoke(Request $request, Response $response, callable $out = null)
```
Comment (lboynton) [inline src/App/Middleware/Authentication.php]: RTH

File: templates/app/account/login.phtml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+$this->headTitle($title);
+?>
+
+</div><!-- end of gradient-wrapper -->
+
+<div class="wrapper content">
+    <h1><?php echo $title; ?></h1>
+
+    <?php
+    if (null !== $error) {
```
Comment (lboynton) [inline templates/app/account/login.phtml]: Do you prefer this style over `<?php if (null !== $error): ?>`?

File: test/AppTest/Service/Authentication/ZendAuthenticationServiceTest.php [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+        self::assertTrue(
+            (new ZendAuthenticationService($users, $storage))->authenticate(
+                'foo@bar.com',
+                'correct horse battery staple'
+            )
+        );
+    }
+
+    public function testHasIdentityReturnsTrueWhenUserIsAuthenticated()
```
Comment (lboynton) [inline test/AppTest/Service/Authentication/ZendAuthenticationServiceTest.php]: This is named differently to what its testing, and so is the next method :)

File: templates/app/account/login.phtml (hunk 6 above)
Comment (lboynton) [inline templates/app/account/login.phtml]: Also `$error` isn't being passed in to the view at all, so it's undefined

**Comment by asgrim** [comment]
@lboynton all comments addressed - happy to merge if build passes?

**Comment by lboynton** [comment]
Yep

**Pull request merged**